ibm/resource_ibm_network_gateway.go

```diff
@@ -0,0 +1,996 @@
+	if !*result.PrivateNetworkOnlyFlag && result.PrimaryIpAddress != nil {
+		connInfo["host"] = *result.PrimaryIpAddress
+	} else {
+		connInfo["host"] = *result.PrimaryBackendIpAddress
+	}
+	d.SetConnInfo(connInfo)
+
+	return nil
+}
+func resourceIBMNetworkGatewayUpdate(d *schema.ResourceData, meta interface{}) error {
+	id, _ := strconv.Atoi(d.Id())
+	service := services.GetHardwareService(meta.(ClientSession).SoftLayerSession())
+
+	if d.HasChange("tags") {
+		err := setHardwareTags(id, d, meta)
+		if err != nil {
+			return err
+		}
+	}
+
+	if d.HasChange("notes") {
+		err := setHardwareNotes(id, d, meta)
+		if err != nil {
+			return err
+		}
+	}
+	err := modifyStorageAccess(service.Id(id), id, meta, d)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceIBMNetworkGatewayDelete(d *schema.ResourceData, meta interface{}) error {
+	sess := meta.(ClientSession).SoftLayerSession()
+	service := services.GetHardwareService(sess)
+
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return fmt.Errorf("Not a valid ID, must be an integer: %s", err)
+	}
+
+	_, err = waitForNoBareMetalActiveTransactions(id, meta)
+	if err != nil {
+		return fmt.Errorf("Error deleting Network Gateway while waiting for zero active transactions: %s", err)
+	}
+
+	billingItem, err := service.Id(id).GetBillingItem()
+	if err != nil {
+		return fmt.Errorf("Error getting billing item for Network Gateway: %s", err)
+	}
+
+	// Monthly  Softlayer items only support an anniversary date cancellation option.
+	billingItemService := services.GetBillingItemService(sess)
+	_, err = billingItemService.Id(*billingItem.Id).CancelItem(
+		sl.Bool(d.Get("hourly_billing").(bool)), sl.Bool(true), sl.String("No longer required"), sl.String("Please cancel this Network Gateway"),
+	)
+	if err != nil {
+		return fmt.Errorf("Error canceling the Network Gateway (%d): %s", id, err)
+	}
+
+	return nil
+}
+
+func resourceIBMNetworkGatewayExists(d *schema.ResourceData, meta interface{}) (bool, error) {
+	service := services.GetHardwareService(meta.(ClientSession).SoftLayerSession())
+
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return false, fmt.Errorf("Not a valid ID, must be an integer: %s", err)
+	}
+
+	result, err := service.Id(id).GetObject()
+	if err != nil {
+		if apiErr, ok := err.(sl.Error); !ok || apiErr.StatusCode != 404 {
+			return false, fmt.Errorf("Error trying to retrieve Network Gateway: %s", err)
+		}
+	}
+
+	return result.Id != nil && *result.Id == id, nil
+}
+
+func getMonthlyGatewayOrder(d *schema.ResourceData, meta interface{}) (datatypes.Container_Product_Order, error) {
+
+	d.Set("package_key_name", "NETWORK_GATEWAY_APPLIANCE")
+	d.Set("os_reference_code", "OS_VYATTA_5600_5_X_UP_TO_1GBPS_SUBSCRIPTION_EDITION_64_BIT")
+	d.Set("process_key_name", "INTEL_SINGLE_XEON_1270_3_40_2")
+	d.Set("os_key_name", "OS_VYATTA_5600_5_X_UP_TO_1GBPS_SUBSCRIPTION_EDITION_64_BIT")
+	d.Set("public_bandwidth", "BANDWIDTH_20000_GB")
```

Comment:
Shouldn't thes come from user input rather than being hard-coded. As package_key_name is always going to have `NETWORK_GATEWAY_APPLIANCE` as value can we we drop
```
+			// Monthly only
+			"package_key_name": {
+				Type:             schema.TypeString,
+				Optional:         true,
+				ForceNew:         true,
+				DiffSuppressFunc: applyOnce,
+			},
```

Reply:
Included all values in Default properties, also will document them as variables for variables.tf files